### DISCUSSION TITLE
Response model validation for async path operation functions blocks the event loop

### DISCUSSION THREAD

**Discussion by lwfitzgerald** [fastapi/fastapi#11406] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import time
from typing import Annotated

import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel, AfterValidator

app = FastAPI()


def delay_validator(val: int) -> int:
    # Artificially slow validator for demonstration purposes
    time.sleep(10)
    return val


class Model(BaseModel):
    field: Annotated[int, AfterValidator(delay_validator)]


@app.get("/test", response_model=Model)
async def test() -> dict:
    return {"field": 42}


if __name__ == "__main__":
    uvicorn.run(app)
```

### Description

* Run the above example under a profiler/debugger and `GET` the `/test` endpoint
* Note that `delay_validator`/ the `time.sleep` occurs in the event loop thread
* It looks like the `if`/`else` blocks may be the wrong way around here - https://github.com/tiangolo/fastapi/pull/888/files#diff-4d573079004a9f3d148baa4658e68e82b8a3d1a95d603fee8177daa92cf65c93R67

### Operating System

macOS

### Operating System Details

```
ProductName:		macOS
ProductVersion:		14.4
BuildVersion:		23E214
```

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

Python 3.11.8

### Additional Context

Screenshot from `py-spy` debugger running the included example. It shows validation occurring in the event loop thread.
<img width="1920" alt="image" src="https://github.com/tiangolo/fastapi/assets/90507/fa81450d-78a8-40a7-a0c6-13eb19cd5c51">

**Comment by YuriiMotov** [comment, marked as answer]
I think it's implemented correctly now.
Validators should not be heavy excepting the case when it requires DB calls (validating DB object with relations).
But since you can't use this feature with async connection (you should preload all your data before validation) it's not our case.
And running the response model validation every time in the threadpool will slow down the FastAPI. So, I think it's not a good idea to do that without strong reasons.

Can you imagine real-life (not artificially slow) heavy validation function?
If yes, I think we can try to justify the need for the 'validate_response_model_in_thread' flag for path operation, so that you can override the default behavior.

**Reply by lwfitzgerald** [reply]
Having had a second read of the PR introducing the code, I now agree with you that it's correct.

Regarding adding something like `validate_response_model_in_thread` - In my application I'm replicating an API that produces large responses (no pagination). The model returned in the response has many nested models, but no additional validation beyond the (basic) types of the fields. Validation was slow enough to block the event loop.
